File: apps/research-desk/src/features/agent/artifactNames.ts
/**
 * Derived file names for canvas documents (P9, K1).
 *
 * The name is DISPLAY AND ADDRESS, never a key: identity stays the
 * artifactId, the name is `slug(title) + ".md"` made unique per session
 * with a `-2` suffix in created order. Byte-exact port of the Python
 * reference (`src/inqtrix/agents/artifact_names.py`), pinned by the
 * shared fixture `tests/fixtures/artifact_name_parity.json` — change
 * one side only via the fixture.
 */

export const ARTIFACT_NAME_FALLBACK = 'dokument'

export function artifactSlug(title: string): string {
  // normalize(NFKD) -> toLowerCase -> strip combining marks ->
  // non-alnum runs to '-' -> strip edge dashes -> THEN slice to 48
  // (a slice landing on a dash keeps it, exactly like the Python side).
  const normalized = title
    .normalize('NFKD')
    .toLowerCase()
    .replace(/[\u0300-\u036f]/g, '')
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '')
    .slice(0, 48)
  return normalized || ARTIFACT_NAME_FALLBACK
}

/**
 * Map artifactId -> file name for documents in CREATED order (the
 * session index's `order` contract, oldest first). Collisions get
 * `-2`, `-3`, ... BEFORE the extension (`bericht-2.md`). Renaming an
 * OLDER document can shift a younger namesake's suffix: accepted,
 * because the name is display and the id stays stable (K1 edge).
 */
export function assignArtifactFileNames(
  items: readonly { artifactId: string; title: string }[],
): Record<string, string> {
  const taken = new Set<string>()
  const names: Record<string, string> = {}
  for (const item of items) {
    const base = artifactSlug(item.title)
    let candidate = base
    let suffix = 2
    while (taken.has(candidate)) {
      candidate = `${base}-${suffix}`
      suffix += 1
    }
    taken.add(candidate)
    names[item.artifactId] = `${candidate}.md`
  }
  return names
}

File: src/inqtrix/agents/artifact_names.py
"""Derived file names for canvas documents (P9, K1).

The name is DISPLAY AND ADDRESS, never a key: identity stays the
``artifact_id``, the name is ``slug(title) + ".md"`` made unique per
session with a ``-2`` suffix in created order. The slug is a byte-exact
port of the frontend's NFKD slug (``slugLabel``-family), pinned by the
shared parity fixture ``tests/fixtures/artifact_name_parity.json`` that
both languages consume — change one side only via the fixture.
"""

from __future__ import annotations

import re
import unicodedata
from collections.abc import Iterable

ARTIFACT_NAME_FALLBACK = "dokument"
_COMBINING_MARKS = re.compile("[\\u0300-\\u036f]")
_NON_ALNUM = re.compile(r"[^a-z0-9]+")

# Only real canvas documents carry file names; diagnostics kinds keep
# their plain registry rendering.
NAMED_ARTIFACT_KINDS = ("memo", "deliverable")


def artifact_slug(title: str) -> str:
    """NFKD slug, operation-for-operation the frontend's order.

    normalize(NFKD) -> toLowerCase -> strip combining marks ->
    non-alnum runs to ``-`` -> strip edge dashes -> THEN slice to 48
    (a slice landing on a dash keeps it, exactly like the TS side).
    """
    normalized = unicodedata.normalize("NFKD", title).lower()
    normalized = _COMBINING_MARKS.sub("", normalized)
    normalized = _NON_ALNUM.sub("-", normalized)
    normalized = normalized.strip("-")[:48]
    return normalized or ARTIFACT_NAME_FALLBACK


def assign_artifact_file_names(
    items: Iterable[tuple[str, str]],
) -> dict[str, str]:
    """Map ``artifact_id -> file name`` for documents in CREATED order.

    Collisions get ``-2``, ``-3``, ... BEFORE the extension
    (``bericht-2.md``, never ``bericht.md-2``); suffix order is the
    deterministic created order the caller supplies (oldest first —
    the session listing's contract). Renaming an OLDER document can
    therefore shift a younger namesake's suffix: accepted, because the
    name is display and the id stays stable (documented K1 edge).
    """
    taken: set[str] = set()
    names: dict[str, str] = {}
    for artifact_id, title in items:
        base = artifact_slug(title)
        candidate = base
        suffix = 2
        while candidate in taken:
            candidate = f"{base}-{suffix}"
            suffix += 1
        taken.add(candidate)
        names[artifact_id] = f"{candidate}.md"
    return names
